Shorthands are still using "initial" for longhands that are set implicitly

darinadler opened the pull request:
#### 11adcf008ef8a0287a435298802c45749dd17cc1
<pre>
Shorthands are still using &quot;initial&quot; for longhands that are set implicitly
<a href="https://bugs.webkit.org/show_bug.cgi?id=248913">https://bugs.webkit.org/show_bug.cgi?id=248913</a>
rdar://problem/103092939

Reviewed by Sam Weinig.

This removes all the remaining cases that use &quot;initial&quot; for longhands, moving to initial values.
Also, as a result we need to start serializing some properties based on checking if longhands
have their initial values, not just use the implicit flag. That&apos;s because in the case of layered
shorthands, we can no longer use the implicit flag; there is nothing that stores those in the
layers. Before the special implicit initial value was used to work around that. It&apos;s a good
direction to have done that for layered shorthands and for border, since we plan to do that
universally; in another change soon we will finish the job.

I have additional removal and tidying, including removing the now unused
createImplicitInitialValue function entirely. But since this is already a large change, holding
those for followup.

* LayoutTests/editing/pasteboard/copy-paste-inserts-clearing-div-expected.txt:
Updated for progression where we use the background shorthand, not background-color.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/css/background-clip-text-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Progression where we serialize
background shorthand instead of background-image and background-position longhands.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt:
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt:
* LayoutTests/fast/css/parsing-text-emphasis-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Expect various shorthands to be handled
correctly, such as adding border-image, handling the background shorthand, and others.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt:
Expect test to pass.

* LayoutTests/fast/css/transform-inline-style-expected.txt: Don&apos;t expect initial values &quot;ease&quot; and
&quot;0s&quot; to be serialized.
* LayoutTests/fast/css/transform-inline-style-remove-expected.txt: Ditto, also initial values
&quot;none&quot; and &quot;running&quot;.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/border-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Expect more tests to pass.

* Source/WebCore/css/CSSProperties.json: Added the border-image longhands to the border shorthand,
because the border shorthand sets all these properties to initial values. Also sort the offset
shorthand longhands in the order from the specification.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const): If border-image shorthand values don&apos;t all
have their initial values, then do not serialize the border shorthand.
(WebCore::StyleProperties::get2Values const): Rewrote more simply, removing the isInitialValue
checks since commonShorthandChecks takes care of that now. Also use makeString instead of
StringBuilder since we can and it&apos;s more efficient.
(WebCore::StyleProperties::get4Values const): Rewrote more simply, removing the isInherited
and isInitialValue checks since commonShorthandChecks takes care of that now. Also use makeString
instead of StringBuilder since we can and it&apos;s more efficient.
(WebCore::isValueID): Added.
(WebCore::isPair): Added.
(WebCore::StyleProperties::getLayeredShorthandValue const): Rewrote this to use initial values
instead of relying on implicit flags to know what to not serialize. Added various rules that were
not present in the old version that are now needed for round tripping and omitting any value that
can be without changing the meaning.
(WebCore::StyleProperties::borderImagePropertyValue const): Serialize as &quot;none&quot; when everything
has the implicit flag set.
(WebCore::StyleProperties::borderPropertyValue const): Rewrote for simplicitly. Now checks for
initial values instead of depending on the implicit flag, using string comparisons since that&apos;s
the easiest way to build on getCommonValue. Also use makeString instead of StringBuilder.
(WebCore::StyleProperties::hasAllInitialValues const): Added.
(WebCore::canUseShorthandForLonghand): Allow use of the CSSPropertyBackground shorthand,
which did not work well before getLayeredShorthandValue omitted initial values.

* Source/WebCore/css/StyleProperties.h: Added hasAllInitialValues.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addExpandedProperty): Renamed from addExpandedPropertyForValue
and added an implicit flag.
(WebCore::CSSPropertyParser::addExpandedPropertyWithImplicitDefault): Added.
(WebCore::CSSPropertyParser::parseValueStart): Updated for addExpandedProperty name change.
(WebCore::CSSPropertyParser::consumeCSSWideKeyword): Ditto.
(WebCore::initialValueForLonghand): Added.
(WebCore::initialCSSValueForLonghand): Added.
(WebCore::isNumber): Added.
(WebCore::pairValue): Added.
(WebCore::quadValue): Added.
(WebCore::isValueIDPair): Added.
(WebCore::isNumericQuad): Added.
(WebCore::isInitialValueForLonghand): Added.
(WebCore::initialValueTextForLonghand): Added.
(WebCore::CSSPropertyParser::consumeShorthandGreedily): Use initialCSSValueForLonghand for
implicit defaults rather than the implicit initial value.
(WebCore::CSSPropertyParser::consumeBorderShorthand): Renamed from consumeBorder. Do the calls
to addExpandedPropertyWithImplicitDefault here instead of in the caller, continuing to treat
the defaults as implicit values for now.
(WebCore::CSSPropertyParser::consumeBorderImage): Fixed CSSPropertyWebkitMaskBoxImageRepeat and
CSSPropertyBorderImageRepeat to use coalesced pairs of &quot;stretch&quot; for their initial values
rather than a single &quot;stretch&quot;.
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Use initialCSSValueForLonghand instead
of ComputedStyleExtractor::addValueForAnimationPropertyToList.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Removed local variable &quot;implicit&quot;
that was always false. Use initialCSSValueForLonghand for implicit defaults rather than
implicit initial value.
(WebCore::CSSPropertyParser::parseShorthand): Use consumeBorderShorthand. For CSSPropertyBorder
itself, also set all the border-image longhands to their initial values.

* Source/WebCore/css/parser/CSSPropertyParser.h: Updated the addExpandedProperty functions.
Also added isInitialValueForLonghand and initialValueTextForLonghand, exported by the parser
but used by the serialization code.

Canonical link: <a href="https://commits.webkit.org/258061@main">https://commits.webkit.org/258061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af7a9fb6cfc480ef304c6abacd05cf8d0aafbfd9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100779 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110076 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170351 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10861 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/566 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93180 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107923 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91456 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34815 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22849 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77789 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3615 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24373 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3639 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9750 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43872 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5532 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5416 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -273,25 +273,25 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
         return applyProperty(newId, valueToApply.get(), linkMatchMask);
     }
 
+    auto valueID = is<CSSPrimitiveValue>(valueToApply.get()) ? downcast<CSSPrimitiveValue>(valueToApply.get()).valueID() : CSSValueInvalid;
```
This could use the `WebCore::valueID()` function

darinadler replied:
Sure, will do.

Loirooriol commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2221,6 +2393,17 @@ bool CSSPropertyParser::consumeListStyleShorthand(bool important)
 
 bool CSSPropertyParser::parseShorthand(CSSPropertyID property, bool important)
 {
+    auto parseBorder = [&](CSSPropertyID widthProperty, CSSPropertyID styleProperty, CSSPropertyID colorProperty) {
```
Most shorthands will not use this, so I would move it out into its own method.

darinadler replied:
I’m not sure that matters. I don’t think the local variable for a lambda not called incurs overhead. If I made this a separate function it would need more arguments, but I guess I’ll do that, though.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1096,13 +1045,12 @@ String StyleProperties::borderImagePropertyValue(const StylePropertyShorthand& s
         auto value = getPropertyCSSValue(longhand);
         if (!value)
-            return String();
+            return { };
```
This file has so many inconsistencies between `String()`, `nullString()`, `{ }`, and some places even using `emptyString()`. Choosing one and switching all may be better left for another patch?

darinadler replied:
Sure I will revert this. Recently @anttijk told me we normally use this syntax and I think it stuck in my head.

The null vs. empty string is a more serious topic though, not just syntax.

Loirooriol commented on Source/WebCore/css/DeprecatedCSSOMValue.cpp:
```diff
@@ -85,15 +85,20 @@ unsigned short DeprecatedCSSOMComplexValue::cssValueType() const
-    if (m_value->isRevertValue())
-        return CSS_REVERT;
+    if (auto primitiveValue = dynamicDowncast<CSSPrimitiveValue>(m_value.get())) {
+        switch (primitiveValue->valueID()) {
```
Why not just `switch(valueID(m_value.get()))`?

darinadler replied:
Sure, good suggestion.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
```diff
@@ -9,5 +9,5 @@ PASS All aliases can serialize target property set to its initial value
 PASS Can't serialize shorthand when longhands have different priority
 PASS Can't serialize shorthand set to 'initial' when some longhand is missing
-FAIL Can't serialize shorthand set to initial value when some longhand is missing assert_array_equals: lengths differ, expected array [] length 0, got […, ["font", "line-height", "16px -webkit-standard"], ["font", "font-size-adjust", "16px -webkit-standard"], ["font", "font-kerning", "16px -webkit-standard"], ["font", "font-variant-alternates", "16px -webkit-standard"], ["font", "font-variant-lig ...
+FAIL Can't serialize shorthand set to initial value when some longhand is missing assert_array_equals: lengths differ, expected array [] length 0, got […, ["font", "line-height", "16px -webkit-standard"], ["font", "font-size-adjust", "16px -webkit-standard"], ["font", "font-kerning", "16px -webkit-standard"], ["font", "font-variant-alternates", "16px -webkit-standard"], ["font", "font-variant-lig ...
```
Why is this increasing?

darinadler replied:
After I fixed some more things, this went away. So it was probably a bug.

Loirooriol commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1509,7 +1509,12 @@
+                    "border-image-slice",
+                    "border-image-width",
+                    "border-image-outset",
+                    "border-image-repeat"
```
This is good, but in `StyleProperties::getPropertyValue` you should check that these new longhands are present and set to their initial values before serializing `CSSPropertyBorder`. Otherwise it would be bad for `asTextInternal()`

darinadler replied:
Good point. I should probably add a test too.

darinadler replied:
I decided to put this into `commonShorthandChecks`.

darinadler replied:
I am really close to removing all the shorthand-specific versions of this since it’s now checked in `commonShorthandChecks` for almost all cases. And remove all the exceptions in `commonShorthandChecks`. I’ll do those in a follow-up because this change is getting big.

Loirooriol replied:
Now `commonShorthandChecks` checks that these new longhands are present, but `borderPropertyValue` or `getPropertyValue` should check that they are set to the initial value

darinadler replied:
Yes, that jogs my memory. I was adding support for all the border-image longhands to `isInitialValueForLonghand` and it was a tedious project because of the way they are always parsed into quads even when it’s a single value. I decided to not include that work in this pull request because it was already kind of large.

But I suppose if I don’t do all that work then this will create a regression, where `border` will serialize and the `border-image` value will be lost.

Loirooriol replied:
Maybe just check that `getPropertyValue(CSSPropertyBorderImage)` is `"none"` or something like that?
Or don't add these longhands for now.

darinadler replied:
I fixed it the "hard" way.

Loirooriol reviewed (commented):
Overall it seems very nice, but it's a big patch, I can't look at it in detail now. Just some comments.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-valid.html:
```diff
@@ -22,14 +22,14 @@
 test_valid_value("animation", "reverse", ["reverse", "0s ease 0s 1 reverse none running none"]);
 test_valid_value("animation", "both", ["both", "0s ease 0s 1 normal both running none"]);
 test_valid_value("animation", "paused", ["paused", "0s ease 0s 1 normal none paused none"]);
-test_valid_value("animation", "none", ["none", "0s ease 0s 1 normal none running none"]);
```
I would keep allowing `none`, I don't think it's completely clear what should happen when there are multiple minimal serializations https://github.com/w3c/csswg-drafts/issues/8155

You can just add `0s` as a third possibility.

darinadler replied:
OK. Thank you.

Also, I would like to standardize what should happen when there are multiple minimal serializations. This is the kind of difference that can lead to real world interoperability problems.

darinadler replied:
I don’t personally have a preference for `0s` over `none`, but I thought the rule that the first longhand should be the one serialized in the minimal case was a good one.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-valid.html:
```diff
@@ -22,14 +22,14 @@
   "1s cubic-bezier(0, -2, 1, 3) -3s 4 reverse both paused anim");
 
 test_valid_value("animation", "anim paused both reverse, 4 1s -3s cubic-bezier(0, -2, 1, 3)",
-  "0s ease 0s 1 reverse both paused anim, 1s cubic-bezier(0, -2, 1, 3) -3s 4 normal none running none");
```
Other cases have an array with both a short and a full variants, so for consistency I would do the same here.

darinadler replied:
I see, I didn’t realize that’s how it worked.

darinadler replied:
I think eventually we will want a test that **requires** the minimal serialization; for interoperability we want the browsers to handle such cases consistently. But for now I will follow suit in treating it as optional here.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
```diff
@@ -11,52 +11,50 @@
 // <mask-reference> = none | <image> | <mask-source>
-// <masking-mode> = alpha | luminance | auto
 test_valid_value('mask', 'none');
-test_valid_value('mask', 'none alpha');
```
Why remove this?

darinadler replied:
I didn’t exactly **remove** it. I simply changed the test to just specify `alpha` rather than `none alpha` and moved it down below to the `<masking-mode>` section. When the test was originally written, the `<mask-reference>` was mandatory in the specification and couldn’t be omitted. There would be no problem including a `none alpha` as well as an `alpha` test, but there really weren’t too many of those here, with initial values intended to be omitted, so I didn’t think it should be added. Happy to add it if you like.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
```diff
@@ -11,52 +11,50 @@
 test_valid_value('mask', 'linear-gradient(to left bottom, red, blue)');
 test_valid_value('mask', 'linear-gradient(to left bottom, red, blue) luminance');
 test_valid_value('mask', 'url("https://{{host}}/")');
-test_valid_value('mask', 'url("https://{{host}}/") alpha');
```
Ditto

darinadler replied:
This test was moved, not removed. It’s below in the `<masking-mode>` section. I reorganized this a little bit.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values.html:
```diff
@@ -262,10 +262,11 @@
-                     'top', 'center', 'bottom']],
+                     'top', 'bottom']],
                    'inherit'],
+        //FIXME testing cases involving center requires understanding when it can be omitted
```
I'm afraid nobody will address this, so I think you should do it now.

darinadler replied:
There are many, many cases that this test doesn’t currently cover. Many properties just mentioned and commented out. To me, this one didn’t really stick out. Addressing this is a significant project because this test doesn’t cover any cases of omitting values from shorthands at all and is written in a very simple way. I will do it, but I was worried about making this large change a gigantic project. I could do all the WPT fixes separately first without any changes to WebKit, I guess, then this would be more manageable.

darinadler replied:
OK, I did this.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -151,14 +151,9 @@ String StyleProperties::commonShorthandChecks(const StylePropertyShorthand& shor
     std::optional<bool> importance;
     size_t numSetFromShorthand = 0;
     for (auto longhand : shorthand) {
-        // FIXME: Shouldn't serialize the shorthand if some longhand is missing.
```
I think it would be good to try to split the patch into smaller ones, for example I believe this fix could be done first by itself.

darinadler replied:
Sure I’d be willing to split into smaller ones. The biggest job is rebasing all the regression tests, which makes splitting it up a long process; that’s what’s taken me so long here. I am a little worried that some of the changes break things that are then fixed by other changes, so splitting may be a challenging project.

darinadler replied:
I have greatly reduced the patch.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
+                auto areEqual = [](const RefPtr<CSSValue>& a, const RefPtr<CSSValue>& b) {
+                    return a && b && a->equals(*b);
+                };
+                if (areEqual(layerValues[j - 1], layerValues[j])) {
```
This makes me nervous. What is preventing this line from being called on the first iteration of the loop, when j is 0? Just that CSSPropertyBackgroundClip, CSSPropertyMaskClip and CSSPropertyWebkitMaskClip are not the first longhand? Adding an ASSERT would help.

darinadler replied:
I added lots of assertions. It would be nice to write this in another style.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
+            // A single geometry-box value sets both mask-origin and -webkit-mask-clip.
+            if (longhand == CSSPropertyBackgroundClip || longhand == CSSPropertyMaskClip || longhand == CSSPropertyWebkitMaskClip) {
+                // The previous property is origin.
+                auto areEqual = [](const RefPtr<CSSValue>& a, const RefPtr<CSSValue>& b) {
```
Is this different than compareCSSValuePtr?

darinadler replied:
It is. That function treats two null pointers as equal. But I can use it here I think.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
+                    if (i)
+                        skipSerializing = true;
+                } else {
+                    // Color is only in the last layer.
```
Anyway to statically assert this? And if not currently, is there some metadata we should consider exposing from the generators to make this statically assertable?

darinadler replied:
This is not new. I wrote the code in an easier to read way, but the logic was there before. I’m not sure there’s anything to assert really. I’m not going to change anything here.

darinadler replied:
Maybe you can figure out a way to improve this later, but I was really just trying to preserve what was there already in a style that was easy enough to read.

darinadler replied:
This rule about what is in what layer comes from the specification text and is not present in the BNF as far as I can tell.

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
+    Vector<Ref<CSSValue>> values;
+    values.reserveInitialCapacity(length);
+    size_t numLayers = 1;
+    for (unsigned i = 0; i < length; ++i) {
```
I think this loop can be a range-based loop as `I` is only used in the one place.

weinig commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1175,6 +1183,180 @@ bool CSSPropertyParser::consumeColumns(bool important)
+static constexpr InitialValue initialValueForLonghand(CSSPropertyID longhand)
+{
+    // Currently, this tries to cover just longhands that can be omitted from shorthands when parsing or serializing.
+    // Later, we likely want to cover all properties, and generate the table from CSSProperties.json.
```
Nice.

darinadler committed:
Shorthands are still using "initial" for longhands that are set implicitly
https://bugs.webkit.org/show_bug.cgi?id=248913
rdar://problem/103092939

Reviewed by Sam Weinig.

This removes all the remaining cases that use "initial" for longhands, moving to initial values.
Also, as a result we need to start serializing some properties based on checking if longhands
have their initial values, not just use the implicit flag. That's because in the case of layered
shorthands, we can no longer use the implicit flag; there is nothing that stores those in the
layers. Before the special implicit initial value was used to work around that. It's a good
direction to have done that for layered shorthands and for border, since we plan to do that
universally; in another change soon we will finish the job.

I have additional removal and tidying, including removing the now unused
createImplicitInitialValue function entirely. But since this is already a large change, holding
those for followup.

* LayoutTests/editing/pasteboard/copy-paste-inserts-clearing-div-expected.txt:
Updated for progression where we use the background shorthand, not background-color.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/css/background-clip-text-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/background-position-serialize-expected.txt: Progression where we serialize
background shorthand instead of background-image and background-position longhands.

* LayoutTests/fast/css/border-shorthand-initialize-longhands-expected.txt:
* LayoutTests/fast/css/longhand-overrides-shorthand-prefixing-expected.txt:
* LayoutTests/fast/css/parsing-text-emphasis-expected.txt:
Expect more tests to pass.

* LayoutTests/fast/css/remove-shorthand-expected.txt: Expect various shorthands to be handled
correctly, such as adding border-image, handling the background shorthand, and others.

* LayoutTests/fast/css/shorthand-mismatched-list-crash-expected.txt:
Expect test to pass.

* LayoutTests/fast/css/transform-inline-style-expected.txt: Don't expect initial values "ease" and
"0s" to be serialized.
* LayoutTests/fast/css/transform-inline-style-remove-expected.txt: Ditto, also initial values
"none" and "running".

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-position-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-multicol/parsing/column-rule-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-cssText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/border-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-values-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/shorthand-values-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Expect more tests to pass.

* Source/WebCore/css/CSSProperties.json: Added the border-image longhands to the border shorthand,
because the border shorthand sets all these properties to initial values. Also sort the offset
shorthand longhands in the order from the specification.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const): If border-image shorthand values don't all
have their initial values, then do not serialize the border shorthand.
(WebCore::StyleProperties::get2Values const): Rewrote more simply, removing the isInitialValue
checks since commonShorthandChecks takes care of that now. Also use makeString instead of
StringBuilder since we can and it's more efficient.
(WebCore::StyleProperties::get4Values const): Rewrote more simply, removing the isInherited
and isInitialValue checks since commonShorthandChecks takes care of that now. Also use makeString
instead of StringBuilder since we can and it's more efficient.
(WebCore::isValueID): Added.
(WebCore::isPair): Added.
(WebCore::StyleProperties::getLayeredShorthandValue const): Rewrote this to use initial values
instead of relying on implicit flags to know what to not serialize. Added various rules that were
not present in the old version that are now needed for round tripping and omitting any value that
can be without changing the meaning.
(WebCore::StyleProperties::borderImagePropertyValue const): Serialize as "none" when everything
has the implicit flag set.
(WebCore::StyleProperties::borderPropertyValue const): Rewrote for simplicitly. Now checks for
initial values instead of depending on the implicit flag, using string comparisons since that's
the easiest way to build on getCommonValue. Also use makeString instead of StringBuilder.
(WebCore::StyleProperties::hasAllInitialValues const): Added.
(WebCore::canUseShorthandForLonghand): Allow use of the CSSPropertyBackground shorthand,
which did not work well before getLayeredShorthandValue omitted initial values.

* Source/WebCore/css/StyleProperties.h: Added hasAllInitialValues.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::addExpandedProperty): Renamed from addExpandedPropertyForValue
and added an implicit flag.
(WebCore::CSSPropertyParser::addExpandedPropertyWithImplicitDefault): Added.
(WebCore::CSSPropertyParser::parseValueStart): Updated for addExpandedProperty name change.
(WebCore::CSSPropertyParser::consumeCSSWideKeyword): Ditto.
(WebCore::initialValueForLonghand): Added.
(WebCore::initialCSSValueForLonghand): Added.
(WebCore::isNumber): Added.
(WebCore::pairValue): Added.
(WebCore::quadValue): Added.
(WebCore::isValueIDPair): Added.
(WebCore::isNumericQuad): Added.
(WebCore::isInitialValueForLonghand): Added.
(WebCore::initialValueTextForLonghand): Added.
(WebCore::CSSPropertyParser::consumeShorthandGreedily): Use initialCSSValueForLonghand for
implicit defaults rather than the implicit initial value.
(WebCore::CSSPropertyParser::consumeBorderShorthand): Renamed from consumeBorder. Do the calls
to addExpandedPropertyWithImplicitDefault here instead of in the caller, continuing to treat
the defaults as implicit values for now.
(WebCore::CSSPropertyParser::consumeBorderImage): Fixed CSSPropertyWebkitMaskBoxImageRepeat and
CSSPropertyBorderImageRepeat to use coalesced pairs of "stretch" for their initial values
rather than a single "stretch".
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Use initialCSSValueForLonghand instead
of ComputedStyleExtractor::addValueForAnimationPropertyToList.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Removed local variable "implicit"
that was always false. Use initialCSSValueForLonghand for implicit defaults rather than
implicit initial value.
(WebCore::CSSPropertyParser::parseShorthand): Use consumeBorderShorthand. For CSSPropertyBorder
itself, also set all the border-image longhands to their initial values.

* Source/WebCore/css/parser/CSSPropertyParser.h: Updated the addExpandedProperty functions.
Also added isInitialValueForLonghand and initialValueTextForLonghand, exported by the parser
but used by the serialization code.

Canonical link: https://commits.webkit.org/258061@main
